cpp/include/raft/matrix/detail/scatter.cuh

```diff
@@ -0,0 +1,127 @@
+ * @param[inout] map map containing the destination index for each row (n_rows)
+ * @param[inout] batch_size column batch size
+ */
+
+template <typename MatrixT, typename IndexT>
+void scatterInplaceImpl(
+  raft::resources const& handle,
+  raft::device_matrix_view<MatrixT, IndexT, raft::layout_c_contiguous> inout,
+  raft::device_vector_view<const IndexT, IndexT, raft::layout_c_contiguous> map,
+  IndexT batch_size)
+{
+  IndexT m          = inout.extent(0);
+  IndexT n          = inout.extent(1);
+  IndexT map_length = map.extent(0);
+
+  // skip in case of 0 length input
+  if (map_length <= 0 || m <= 0 || n <= 0 || batch_size < 0) return;
+
+  RAFT_EXPECTS(map_length == m,
+               "Length of map should be equal to number of rows for inplace scatter");
+
+  // re-assign batch_size for default case
+  if (batch_size == 0) batch_size = n;
+
+  RAFT_EXPECTS(batch_size <= n, "batch size should be <= number of columns");
+
+  auto exec_policy = resource::get_thrust_policy(handle);
+
+  IndexT n_batches = raft::ceildiv(n, batch_size);
+
+  auto scratch_space = raft::make_device_vector<MatrixT, IndexT>(handle, m * batch_size);
+
+  for (IndexT bid = 0; bid < n_batches; bid++) {
+    IndexT batch_offset   = bid * batch_size;
+    IndexT cols_per_batch = min(batch_size, n - batch_offset);
+
+    auto copy_op = [inout = inout.data_handle(),
+                    map   = map.data_handle(),
+                    batch_offset,
+                    cols_per_batch = raft::util::FastIntDiv(cols_per_batch),
+                    n] __device__(auto idx) {
+      IndexT row = idx / cols_per_batch;
+      IndexT col = idx % cols_per_batch;
+      return inout[row * n + batch_offset + col];
+    };
+    raft::linalg::map_offset(
+      handle,
+      raft::make_device_vector_view(scratch_space.data_handle(), m * cols_per_batch),
+      copy_op);
+
+    auto scatter_op = [inout         = inout.data_handle(),
+                       map           = map.data_handle(),
+                       scratch_space = scratch_space.data_handle(),
+                       batch_offset,
+                       cols_per_batch = raft::util::FastIntDiv(cols_per_batch),
+                       n] __device__(auto idx) {
+      IndexT row     = idx / cols_per_batch;
+      IndexT col     = idx % cols_per_batch;
+      IndexT map_val = map[row];
+
+      inout[map_val * n + batch_offset + col] = scratch_space[idx];
+      return;
+    };
+    auto counting = thrust::make_counting_iterator<IndexT>(0);
+    thrust::for_each(exec_policy, counting, counting + m * cols_per_batch, scatter_op);
+  }
+}
+
+template <typename MatrixT, typename IndexT>
+void scatter(raft::resources const& handle,
+             raft::device_matrix_view<MatrixT, IndexT, raft::layout_c_contiguous> inout,
+             raft::device_vector_view<const IndexT, IndexT, raft::layout_c_contiguous> map,
+             IndexT batch_size)
+{
+  scatterInplaceImpl(handle, inout, map, batch_size);
+}
```

Comment:
That is because if the scatter map is not a perfect permutation of the range `[0, n_rows)`, it will lead to a data race. I thought that demanding that the output of `transformOp` satisfies this condition would demand too much of the transformOp. Users can always use `map_offset` or `thrust` to create a transformed map beforehand if need be. Another thing to consider is that since this is a hard requirement on the map, it might make sense to completely get rid of scatter and simply use gather with the inverse of the map. This just has the additional cost of creating the inverse map.

Reply:
Good point :+1:. Let's leave it as is.